src/fheroes2/dialog/dialog_system_options.cpp

```diff
@@ -214,129 +208,140 @@ namespace
-        const fheroes2::Rect & heroSpeedRoi = roi[3];
-        const fheroes2::Rect & aiSpeedRoi = roi[4];
-        const fheroes2::Rect & scrollSpeedRoi = roi[5];
-        const fheroes2::Rect & interfaceTypeRoi = roi[6];
-        const fheroes2::Rect & interfaceStateRoi = roi[7];
-        const fheroes2::Rect & battleResolveRoi = roi[8];
-
-        drawDialog( roi );
-
-        const fheroes2::Point buttonOffset( 112 + dialogArea.x, 362 + dialogArea.y );
-        fheroes2::Button buttonOkay( buttonOffset.x, buttonOffset.y, isEvilInterface ? ICN::BUTTON_SMALL_OKAY_EVIL : ICN::BUTTON_SMALL_OKAY_GOOD, 0, 1 );
-        buttonOkay.draw();
+        fheroes2::ImageRestorer restorer( display, shadowOffset.x, shadowOffset.y, dialog.width() + BORDERWIDTH, dialog.height() + BORDERWIDTH );
+        const fheroes2::Rect windowRoi{ dialogOffset.x, dialogOffset.y, dialog.width(), dialog.height() };
+
+        fheroes2::Blit( dialogShadow, display, windowRoi.x - BORDERWIDTH, windowRoi.y + BORDERWIDTH );
+        fheroes2::Blit( dialog, display, windowRoi.x, windowRoi.y );
+
+        fheroes2::ImageRestorer emptyDialogRestorer( display, windowRoi.x, windowRoi.y, windowRoi.width, windowRoi.height );
+
+        const fheroes2::Rect windowLanguageRoi( languageRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowGraphicsRoi( graphicsRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowAudioRoi( audioRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowHeroSpeedRoi( heroSpeedRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowEnemySpeedRoi( enemySpeedRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowHotKeyRoi( hotKeyRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowInterfaceRoi( interfaceRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowTextSupportModeRoi( textSupportModeRoi + windowRoi.getPosition() );
+        const fheroes2::Rect windowBattlesRoi( battlesRoi + windowRoi.getPosition() );
+
+        const auto drawOptions = [&windowLanguageRoi, &windowGraphicsRoi, &windowAudioRoi, &windowHeroSpeedRoi, &windowEnemySpeedRoi, &windowHotKeyRoi,
+                                  &windowInterfaceRoi, &windowTextSupportModeRoi, &windowBattlesRoi]() {
+            drawLanguage( windowLanguageRoi );
+            drawGraphics( windowGraphicsRoi );
+            drawAudioOptions( windowAudioRoi );
+            drawHeroSpeed( windowHeroSpeedRoi );
+            drawEnemySpeed( windowEnemySpeedRoi );
+            drawHotKeyOptions( windowHotKeyRoi );
+            drawInterfaceSettings( windowInterfaceRoi );
+            drawTextSupportModeOptions( windowTextSupportModeRoi );
+            drawBattles( windowBattlesRoi );
+        };
+
+        drawOptions();
+
+        const auto refreshWindow = [&drawOptions, &emptyDialogRestorer, &display]() {
+            emptyDialogRestorer.restore();
+            drawOptions();
+            display.render();
```

Comment:
The same comment as above.
```suggestion
            display.render( emptyDialogRestorer.rect() );
```